kong/clustering/services/sync/hooks.lua

```diff
@@ -174,6 +180,21 @@ function _M:register_dao_hooks()
     ["dao:upsert:pre"]  = pre_hook_func,
     ["dao:upsert:fail"] = fail_hook_func,
     ["dao:upsert:post"] = post_hook_writer_func,
+
+    -- dao:upsert_by
+    ["dao:upsert_by:pre"]  = pre_hook_func,
+    ["dao:upsert_by:fail"] = fail_hook_func,
```

Comment:
We should also add these new hooks `dao:upsert_by:fail` `dao:delete_by:fail` `dao:update_by:fail` in dao module.